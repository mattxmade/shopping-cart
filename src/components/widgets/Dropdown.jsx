import PropTypes from "prop-types";
import React, { useEffect, useRef, useState } from "react";

import "./Dropdown.style.scss";
import AwesomeSvg from "../svg-icons/Awesome.module";
import LocalStorage from "../../data/LocalStorage.module";

const Dropdown = ({ children, handleSortProductsBy }) => {
  const dropdownRef = useRef();
  const restoreSelection = LocalStorage.get("sort");

  useEffect(() => {
    window.addEventListener("click", closeDropdown);

    if (restoreSelection) {
      setSelection(restoreSelection);
      setTimeout(() => handleSortProductsBy(restoreSelection), 0);
    }

    return () => window.removeEventListener("click", closeDropdown);
  }, []);

  const toggleRef = useRef();
  const [toggleDropdown, setToggleDropdown] = useState(false);
  const [selection, setSelection] = useState(children[0]);

  const closeDropdown = () => {
    if (toggleRef.current) setToggleDropdown(false);
  };

  const sortOptions = children;
  const CaretDown = AwesomeSvg.CaretDownIcon;

  const dropdownStyle = {
    dropdown: {
      zIndex: 2,
      display: "flex",
      width: "100%",
      flexDirection: "column",
      position: "absolute",
      backgroundColor: "white",
      left: 0,
      top: 0,
    },

    option: {
      width: "100%",
      cursor: "pointer",
    },
  };

  const handleSortOption = (string) => {
    dropdownRef.current.focus();
    if (string === selection) return setToggleDropdown(false);

    setSelection(string);
    setToggleDropdown(false);
    handleSortProductsBy(string.toLowerCase());
  };

  useEffect(() => {
    toggleRef.current = toggleDropdown;
    dropdownRef.current.setAttribute("aria-expanded", toggleDropdown);
  }, [toggleDropdown]);

  return (
    <ul
      ref={dropdownRef}
      tabIndex={-1}
      style={dropdownStyle.dropdown}
      className="dropdown"
    >
      <li
        tabIndex={0}
        role="button"
        aria-label="Sort products dropdown menu button"
        style={dropdownStyle.option}
        className="dropdown-selected"
        onClick={() => setTimeout(() => setToggleDropdown(!toggleDropdown), 0)}
        onKeyDown={(e) =>
          e.code === "Space" || e.code === "Enter" || e.code === "NumpadEnter"
            ? setToggleDropdown(!toggleDropdown)
            : null
        }
      >
        {selection} <CaretDown parent={"dropdown"} />
      </li>

      {toggleDropdown ? (
        sortOptions.map((option, index) => (
          <li
            key={index}
            className="dropdown-choice"
            style={dropdownStyle.option}
            onClick={(e) => {
              handleSortOption(e.target.textContent);
            }}
            onKeyDown={(e) =>
              e.code === "Space" || e.code === "Enter" || e.code === "Return"
                ? handleSortOption(e.target.textContent)
                : null
            }
            tabIndex={0}
            role="button"
            aria-label={`Sort by: ${option}`}
          >
            {option}
          </li>
        ))
      ) : (
        <li style={{ width: 0, height: 0 }}></li>
      )}
    </ul>
  );
};

Dropdown.propTypes = {
  children: PropTypes.node,
  handleSetSort: PropTypes.func,
  handleSortProductsBy: PropTypes.func,
};

export default Dropdown;
