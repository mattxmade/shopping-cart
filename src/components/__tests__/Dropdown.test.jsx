import React from "react";
import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";

import "@testing-library/jest-dom";
import "@testing-library/dom";

import Dropdown from "../widgets/Dropdown";

// arrange act assert
const handleSetSort = jest.fn;
const handleSortProductsBy = jest.fn;

const props = {
  children: ["name: a-z", "price: low to high", "price: high to low"],
  handleSetSort,
  handleSortProductsBy,
};

const user = userEvent.setup();

describe("Dropdown Component unit tests", () => {
  it("Renders Dropdown", () => {
    const { asFragment } = render(<Dropdown {...props}></Dropdown>);
    expect(asFragment).toMatchSnapshot();
  });

  it("Renders only primary list item on mount", () => {
    render(<Dropdown {...props} />);
    const listItems = screen.getAllByRole("listitem");

    expect(listItems.length).toEqual(1);

    listItems.map(
      (listItem, index) =>
        expect(listItem.textContent.trim()).toBe(props.children[index]) // [1]
    );
  });

  it("Renders expected list items on dropdown toggle", async () => {
    render(<Dropdown {...props} />);
    const toggleDropdown = screen.getByRole("listitem");

    await user.click(toggleDropdown);
    const listItems = await screen.findAllByRole("listitem");

    expect(listItems.length).toEqual(props.children.length + 1);
  });

  it("On click, renders dropdown option as primary list item", async () => {
    render(<Dropdown {...props} />);
    const toggleDropdown = screen.getByRole("listitem");

    await user.click(toggleDropdown);
    const dropdownOptionsOpen = await screen.findAllByRole("listitem");

    const optionFromDropdown = await dropdownOptionsOpen[
      Math.floor(Math.random() * dropdownOptionsOpen.length)
    ];

    await user.click(optionFromDropdown);
    const dropdownOptionsClosed = await screen.findAllByRole("listitem");

    expect(dropdownOptionsClosed.length).toEqual(1);
    expect(dropdownOptionsClosed[0].textContent.trim()).toBe(
      optionFromDropdown.textContent
    );
  });
});

// [1] returns textContent with whitespace at the end
