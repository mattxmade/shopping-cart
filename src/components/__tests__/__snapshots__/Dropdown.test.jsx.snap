// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component unit tests Renders Dropdown 1`] = `[Function]`;
